03_Modules/Certificates/src/module/service/CertificateAuthority.ts

```diff
@@ -0,0 +1,102 @@
+// Copyright Contributors to the CitrineOS Project
+//
+// SPDX-License-Identifier: Apache 2.0
+
+import {
+  CertificateSigningUseEnumType,
+  ICache,
+  SystemConfig,
+} from '@citrineos/base';
+import { ICertificateAuthorityClient } from '../client/interface';
+import { Hubject } from '../client/hubject';
+import * as forge from 'node-forge';
+import { Acme } from '../client/acme';
+import { ILogObj, Logger } from 'tslog';
+import { Local } from '../client/local';
+
+export class CertificateAuthorityService {
+  private readonly _hubjectClient: ICertificateAuthorityClient;
+  private readonly _acmeClient: ICertificateAuthorityClient;
+  private readonly _localCertificateAuthority: ICertificateAuthorityClient;
+  private readonly _logger: Logger<ILogObj>;
+
+  constructor(config: SystemConfig, cache: ICache, logger?: Logger<ILogObj>) {
+    this._logger = logger
+      ? logger.getSubLogger({ name: this.constructor.name })
+      : new Logger<ILogObj>({ name: this.constructor.name });
+
+    this._acmeClient = new Acme(config, this._logger);
+    this._hubjectClient = new Hubject(config);
+    this._localCertificateAuthority = new Local(config, cache, this._logger);
```

Comment:
I'm a little confused by this section.
If I'm following, we have a 'hubjectClient' which is for V2G Certs, and then 'localCA' or 'acmeClient' for ChargingStation Certs
but here you instantiate both acmeClient and localCA. From what i can see here, acmeClient is effectively unused.

Would the way this works be: 2 different kinds of CA clients (V2G & ChargingStation) with 1 implementation per kind depending on the SystemConfig (so, _either_ acme _or_ local, not both)?

Reply:
> I'm a little confused by this section. If I'm following, we have a 'hubjectClient' which is for V2G Certs, and then 'localCA' or 'acmeClient' for ChargingStation Certs but here you instantiate both acmeClient and localCA. From what i can see here, acmeClient is effectively unused.
> 
> Would the way this works be: 2 different kinds of CA clients (V2G & ChargingStation) with 1 implementation per kind depending on the SystemConfig (so, _either_ acme _or_ local, not both)?

Hi Thana, sorry for this mess. 

The local client is used for signing charging station certificate. It uses our mtls private key (sub CA key) to sign the cert.

The acme client is used for the PR 3 which adds a generate cert endpoint. In that PR 3, the acme client is used to generate the Sub CA certificate signed by Let's encrypt. I initiate it in the CertificateAuthority as well so that api or module script can always only interact with CertificateAuthority rather than multiple clients.

Regarding the last question, for charging station certificate, we can only use local client because we need to use our own Sub CA certificate to sign it due to the common name limitation (unique serial number of charger). For V2G certificate, I am not sure whether it can be signed by acme client(Let's encrypt) as well. Or only hubject can sign and return the correct V2G root cert?

Reply:
Besides, I am thinking about removing the local client and move its functions into the `CertificateAuthority`. Because it is not a real client like hubject and acme which stands for an external service. (I though it might be better organized when I initially added this local client.) And during working for the PR 2 (update tls certs endpoint), I also found the local client brings additional difficulty when updating the content inside the `_securityCaCertsKeyMap`.